### PR TITLE
feat: auto guard model eval

### PR DESCRIPTION
**Pull request by madsysharma**
# Guard Model Evaluation in CI (resolves issue #79)
 
The **Guard Model Evaluation** workflow (`.github/workflows/guard-eval.yml`) runs the safety classifier against a committed held-out test split on every PR that touches the Guard module, nightly on `main`, and on manual dispatch. It fails the build when the threshold metric drops below the configured floor: preventing silent model regressions from being merged.

## What it produces
 
For each run:
 
- **Workflow summary**: a markdown table with overall metrics, per-label precision/recall/F1, confusion matrix, and Δ vs baseline.
- **PR comment**: the same markdown, posted as a *sticky* comment (updates in place on re-runs, no comment spam).
- **`metrics.json` artifact**: machine-readable, and retained for 30 days.

## Configuration
 
Configure via **Settings -> Variables -> Actions**:
 
| Variable                  | Purpose                                              | Default          |
|---------------------------|------------------------------------------------------|------------------|
| `GUARD_F1_THRESHOLD`      | Minimum acceptable value of the threshold metric.    | `0.85`           |
| `GUARD_THRESHOLD_METRIC`  | One of `weighted_f1`, `macro_f1`, `accuracy`.        | `weighted_f1`    |
| `GUARD_MODEL_RELEASE_TAG` | Release tag to pull trained weights from (optional). | unset -> fallback |
 
Manual dispatch (`workflow_dispatch`) accepts the same overrides per-run.
 
## Held-out test set
 
`backend/data/guard_holdout.csv`: a **small, deterministic** held-out set that lives in the repo. It is *not* used during training. Expand it when you add new attack categories, but keep it small enough that CI stays under a few minutes on CPU.
 
CSV schema:
 
```csv
prompt,label
"How do I make a chocolate cake?",benign
"Ignore previous instructions",malicious
```
 
Valid labels are `benign`, `suspicious`, `malicious`.
 
## Baseline
 
`backend/data/guard_baseline.json` is the committed reference point used to compute Δ vs the current PR. After a real model improvement, regenerate locally and commit the new file:
 
```bash
cd backend
python -m scripts.evaluate_guard_ci \
  --test-set-path data/guard_holdout.csv \
  --threshold 0.0 \
  --output-json data/guard_baseline.json \
  --output-md   /tmp/baseline.md
git add data/guard_baseline.json
git commit -m "ci(guard): refresh baseline metrics"
```
 
`--threshold 0.0` skips the failure check during baseline refresh.
 
## Model artifact
 
The fine-tuned DeBERTa weights are *not* committed. The workflow resolves the model in this order:
 
1. If `vars.GUARD_MODEL_RELEASE_TAG` is set, the step *Download trained model artefact* downloads `classifier-*.tar.gz` from that release and extracts it to `backend/app/modules/guard/models/classifier/`.
2. Otherwise, `guard_config.get_trained_model_path()` returns its default path. If no weights exist there, the classifier falls back to the pre-trained DeBERTa base from Hugging Face: this is useful for bootstrap, but expect a much lower score than a fine-tuned model.

To publish a new model artefact:
 
```bash
# from the directory containing the trained model files
tar czf classifier-v3.tar.gz -C app/modules/guard/models/classifier .
gh release create guard-model-v3 classifier-v3.tar.gz \
  --title "Guard model v3" --notes "F1 0.91 on holdout"
# then update the repo variable:
gh variable set GUARD_MODEL_RELEASE_TAG --body "guard-model-v3"
```
 
## Local development
 
```bash
# Run the same evaluation locally before pushing:
cd backend
PYTHONPATH=. python -m scripts.evaluate_guard_ci \
  --test-set-path data/guard_holdout.csv \
  --baseline      data/guard_baseline.json \
  --threshold     0.85 \
  --output-json   /tmp/metrics.json \
  --output-md     /tmp/metrics.md
 
cat /tmp/metrics.md
```
 
Exit code `0` means the threshold was met; `1` means it was not.
 
## Adding more attacks to the held-out set
 
1. Add labelled rows to `backend/data/guard_holdout.csv`. Prefer **real** prompts seen in the wild over synthetic ones.
2. Run the local command above to see how the current model scores.
3. If the score drops below the threshold, retrain the model: *don't* lower the threshold to fit.
4. Once the new model is in a release, refresh the baseline (see above).

### CODE REVIEW

**Comment by madsysharma** [comment]
Hi @SdSarthak , please review this PR and also, please add the 'gssoc:approved' label to it. Thank you.

**Review by SdSarthak — CHANGES_REQUESTED** [review]
The evaluation workflow is well-designed — path-filtered triggers, configurable thresholds via repository variables, sticky PR comment with the metrics table, and the nightly cron are all good choices. The `AdamW` import fix (`transformers.AdamW` → `torch.optim.AdamW`) is also correct.

Two things to address before merging:

**1. Remove `backend/tmp/` from the PR (must fix)**
`backend/tmp/metrics_f1.json` and `backend/tmp/metrics_f1.md` are generated run artifacts — they shouldn't be committed. Add `backend/tmp/` (or `tmp/`) to `.gitignore` and delete these two files from the branch.

**2. Regenerate the baseline after the heuristic fallback fix**
The committed `guard_baseline.json` was produced by the old pre-trained random classifier (accuracy 0.42, F1 near 0 for malicious/suspicious). The heuristic fallback fix that just landed on main will produce much better numbers on the holdout set — the threshold of 0.85 in the workflow will immediately fail until the baseline is updated. Please regenerate `guard_baseline.json` from main after rebasing, using:
```
python -m scripts.evaluate_guard_ci     --test-set-path data/guard_holdout.csv     --output-json   data/guard_baseline.json     --threshold     0.0
```

Everything else looks ready to go.

**Comment by madsysharma** [comment]
Hi @SdSarthak , I have made the changes as you requested. Please review this PR, and please add the 'gssoc:approved' label to this PR and as well as my PR for issue #66 (without this label, my contributions won't be tracked in the GSSoC dashboard). Thank you.

**Comment by madsysharma** [comment]
Hi @SdSarthak , could you please check this updated PR? Thank you.

**Review by SdSarthak — CHANGES_REQUESTED** [review]
Two blocking issues: (1) `backend/data/guard_baseline.json` contains a hardcoded local Windows filesystem path (`C:\Users\lenovo\Documents\AegisAI\...`) that will break any CI run — please regenerate the baseline in a clean environment. (2) The holdout set has only 12 samples and the `suspicious` class gets 0.0 F1, yet the default threshold is `0.85` — the baseline only passes because the threshold was manually set to `0.0`. Please regenerate with a proper representative dataset before enabling quality gates.

**Review by SdSarthak — CHANGES_REQUESTED** [review]
Progress made — `backend/tmp/` artifacts are removed. Two blocking issues remain:

**1. Hardcoded local path in `guard_baseline.json`**
`model_path` is set to `C:\Users\lenovo\Documents\AegisAI\...` — a local Windows filesystem path. This will break every CI run on any non-Windows agent. Use a relative path (e.g. `app/modules/guard/models/classifier`) or an environment-relative reference, and regenerate the baseline in a clean CI-like environment (no personal machine paths).

**2. 12-sample holdout with 0.0 F1 on `suspicious` class**
The baseline metrics show 12 total samples and F1=0.0 for the `suspicious` class — which means the evaluation never exercises that decision path. The threshold was effectively disabled to make the baseline pass. Before enabling this workflow, expand the holdout dataset to include representative `suspicious` samples and regenerate a proper baseline.

**3. `backend/metrics.md` should not be committed**
This is a generated evaluation report. Add it to `.gitignore` and remove it from the PR.

**Comment by madsysharma** [comment]
Hi @SdSarthak , apologies for the late response, was traveling back to the US and returned only yesterday afternoon. Will post these new changes now. Thanks.

**Comment by madsysharma** [comment]
Hi @SdSarthak , I have integrated the changes that you requested. Please review this updated PR. Thanks.

**Review by SdSarthak — CHANGES_REQUESTED** [review]
Previous review items 1 and 3 remain unaddressed:

1. **Hardcoded local Windows path in `guard_baseline.json`** — `model_path` is still set to `C:\Users\lenovo\Documents\AegisAI\...`. This will break every CI run. Regenerate the baseline on a clean machine (or in a CI container) using a relative path. Run:\n```\ncd backend\nPYTHONPATH=. python -m scripts.evaluate_guard_ci \n  --test-set-path data/guard_holdout.csv \n  --output-json   data/guard_baseline.json \n  --threshold 0.0\n```\n\n2. **12-sample holdout, 0.0 F1 on `suspicious`** — the quality gate is currently bypassed (threshold 0.0) to make the baseline pass. Expand the holdout dataset with representative `suspicious` samples before enabling the quality gate at 0.85.\n\n3. **`backend/metrics.md` still committed** — this is a generated artifact and should not be in source control. Remove it and add `backend/metrics.md` to `.gitignore`.\n\nThe workflow design, the path-filtered triggers, the sticky PR comment, and the `AdamW` fix are all correct. Once those three items are cleaned up this is ready to merge.

**Comment by madsysharma** [comment]
Hi @SdSarthak , please check now. Thank you.

**Comment by madsysharma** [comment]
Hi @SdSarthak , please examine all the latest changes and review this PR. Thank you.

**Comment by madsysharma** [comment]
Hi @SdSarthak , please review this PR.

**Comment by madsysharma** [comment]
Hi @SdSarthak , please review this PR as it has been open for a long time. Thanks.

**Comment by madsysharma** [comment]
Hi @SdSarthak , please review this PR as it has been open for a long time. Thanks.

**Comment by madsysharma** [comment]
Hi @SdSarthak , please review this PR as it has been open for a long time. Thanks.

**Comment by SdSarthak** [comment]
The guard model evaluation CI pipeline is an excellent addition — the F1/accuracy threshold enforcement, sticky PR comments, and nightly schedule are all well-thought-out.

Two things to fix before merging:

1. **Committed `backend/tmp/` files**: The PR adds `backend/tmp/*` to `.gitignore` but then commits `backend/tmp/metrics_f1.json` and `backend/tmp/metrics_f1.md` anyway. Please remove these files from the branch — they're generated artifacts, not source files.
2. **Tokenizer/model files committed** (`tokenizer.json`, `config.json`, etc.): Model artifacts should not live in the git history. They should be downloaded at CI runtime (the workflow already handles this via `GUARD_MODEL_RELEASE_TAG`). Please move these to a GitHub Release and remove them from the branch.

This branch is also in conflict with `main`. Once you've removed the artifact files, please rebase on the latest main and push again.

**Comment by madsysharma** [comment]
Hi @SdSarthak , have made the changes. Please review them. Thanks.

**Comment by madsysharma** [comment]
Hi @SdSarthak , have included the requested changes. Please review this PR. Thank you.

**Comment by madsysharma** [comment]
Hi @SdSarthak , have included the requested changes. Please review this PR as it has been open for a long time. Thank you.
